packages/fxa-settings/src/lib/channels/firefox.ts

```diff
@@ -78,12 +87,90 @@ export type FxAOAuthLogin = {
   state: string;
 };
 
+interface WebChannelRequest {
+  command: FirefoxCommand;
+  messageId: string;
+  timeoutId?: number;
+  reject(reason?: any): void;
+  resolve(value?: any): void;
+}
+
+interface OutstandingRequestOptions {
+  window: Window;
+  sendTimeoutLength?: number;
+}
+
+/**
+ * OutstandingRequests manages all Firefox WebChannel requests that
+ * the web application is waiting for a response for.
+ *
+ * This is particularly useful for the fxastatus request, where the the
+ * web application is requesting the signed in user data from the browser.
+ */
+class OutstandingRequests {
+  private window_: Window;
+  private sendTimeoutLength: number;
+  private requests: { [key: string]: WebChannelRequest };
+  constructor(options: OutstandingRequestOptions) {
+    this.window_ = options.window;
+    this.requests = {};
+    this.sendTimeoutLength =
+      options.sendTimeoutLength || DEFAULT_SEND_TIMEOUT_LENGTH_MS;
+  }
+
+  add(messageId: string, request: WebChannelRequest) {
+    this.remove(messageId);
+    request.timeoutId = this.window_.setTimeout(() => {
+      request.reject();
+      this.remove(messageId);
+    }, this.sendTimeoutLength);
+    this.requests[messageId] = request;
+  }
+
+  resolve(message: FirefoxMessage) {
+    const outstanding = this.requests[message.messageId];
+    if (outstanding) {
+      outstanding.resolve(message.data);
+      this.remove(message.messageId);
+    }
+  }
+
+  reject(messageId: string, error: FirefoxMessageError) {
+    const outstanding = this.requests[messageId];
+    if (outstanding) {
+      outstanding.reject(error);
+      this.remove(messageId);
+    }
+  }
+
+  remove(messageId: string) {
+    const outstanding = this.requests[messageId];
+    if (outstanding) {
+      this.window_.clearTimeout(outstanding.timeoutId);
+      delete this.requests[messageId];
+    }
+  }
+}
+
+// Suffix to ensure each message has a unique messageId.
+// Every send increments the suffix by 1.
+// A module variable is used instead of an instance variable because
+// more than one channel can exist. Using an instance variable,
+// it's possible for two messages on two channels to have the same
+// messageId, if both channels send a message in the same millisecond.
+// This might not cause any harm in reality, but this avoids
+// that possibility.
+let messageIdSuffix = 0;
```

Comment:
Sorry for not checking this PR before, if you rebase you'll see we recently added this in #16145.